### PR TITLE
Handle multiple stack api versions

### PR DESCRIPTION
**Pull request by silvin-lubecki**
**- What I did**
I integrated the new version of the Stack API (v1beta2). The CLI can now communicates with both versions.

⚠️  This PR has a dependency on #898 ⚠️ 

**- How I did it**
I added a common interface between v1beta1 and v1beta2 stack clients and one common object stack, so the code in the commands remains the same. 

**- How to verify it**
Execute a `docker version` command against latest `docker for mac`, it should output at the end:
```
 Orchestrator:
   Kubernetes:  v1.9.2
   Stack:       v1beta2
```

**- Description for the changelog**
* Add support for the new Stack API for Kubernetes v1beta2

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

File: cli/command/stack/kubernetes/stack.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// createFileBasedConfigMaps creates a Kubernetes ConfigMap for each Compose global file-based config.
+func (s *stack) createFileBasedConfigMaps(configMaps corev1.ConfigMapInterface) error {
+	for name, config := range s.spec.Configs {
+		if config.File == "" {
+			continue
+		}
+
+		fileName := path.Base(config.File)
+		content, err := ioutil.ReadFile(config.File)
+		if err != nil {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stack.go]: - Do we want to *group* errors ?
- If a file can't read (but not the first one), a configMap will be created in k8s and not used (orphan) — **and** will probably cause errors next time we do the `docker stack deploy`, right ?

I think we should "read all", and if there is no error, "create all". Then, if one of the creation fails, what should we do with the previous one created ?

File: cli/command/stack/kubernetes/stack.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+// createFileBasedSecrets creates a Kubernetes Secret for each Compose global file-based secret.
+func (s *stack) createFileBasedSecrets(secrets corev1.SecretInterface) error {
+	for name, secret := range s.spec.Secrets {
+		if secret.File == "" {
+			continue
+		}
+
+		fileName := path.Base(secret.File)
+		content, err := ioutil.ReadFile(secret.File)
+		if err != nil {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stack.go]: Same question here

File: cli/command/stack/kubernetes/stackclient.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	return nil
+}
+
+// stackV1Beta2 implements stackClient interface and talks to compose component v1beta2.
+type stackV1Beta2 struct {
+	stacks composev1beta2.StackInterface
+}
+
+func (c *KubeCli) newStackV1Beta2() (stackClient, error) {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stackclient.go]: I would rather have `func newStackV1Beta2(config *rest.Config, namespace string) …` here (not tying it up with `KubeCli`)

File: cli/command/stack/kubernetes/stackclient.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	List(opts metav1.ListOptions) ([]stack, error)
+	IsColliding(servicesClient corev1.ServiceInterface, s stack) error
+}
+
+// stackV1Beta1 implements stackClient interface and talks to compose component v1beta1.
+type stackV1Beta1 struct {
+	stacks composev1beta1.StackInterface
+}
+
+func (c *KubeCli) newStackV1Beta1() (stackClient, error) {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stackclient.go]: I would rather have `func newStackV1Beta1(config *rest.Config, namespace string) …` here (not tying it up with `KubeCli`)

File: kubernetes/README.md [hunk 5]
```diff
@@ -1,4 +0,0 @@
-# Kubernetes client libraries
```
Comment (vdemeester) [inline kubernetes/README.md]: Can we keep the README ? :angel:

File: kubernetes/client/clientset/clientset.go [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
+	composev1beta1 "github.com/docker/cli/kubernetes/client/clientset/typed/compose/v1beta1"
+	composev1beta2 "github.com/docker/cli/kubernetes/client/clientset/typed/compose/v1beta2"
 	glog "github.com/golang/glog"
 	discovery "k8s.io/client-go/discovery"
 	rest "k8s.io/client-go/rest"
 	flowcontrol "k8s.io/client-go/util/flowcontrol"
 )
 
+// Interface defines the methods a compose kube client should have
+// FIXME(vdemeester) is it required ?
```
Comment (vdemeester) [inline kubernetes/client/clientset/clientset.go]: Ah :sweat_smile: I need to find the answer for it :stuck_out_tongue:

File: kubernetes/client/clientset/clientset.go [hunk 7]
```diff
@@ -30,8 +43,8 @@ func (c *Clientset) ComposeV1beta1() composev1beta1.ComposeV1beta1Interface {
 	return c.ComposeV1beta1Client
 }
 
-// Deprecated: Compose retrieves the default version of ComposeClient.
-// Please explicitly pick a version.
+// Compose retrieves the default version of ComposeClient.
+// deprecated: please explicitly pick a version.
 func (c *Clientset) Compose() composev1beta1.ComposeV1beta1Interface {
```
Comment (vdemeester) [inline kubernetes/client/clientset/clientset.go]: Hum even if it's deprecated, should we return `ComposeV1beta2Client` here instead :angel:

File: kubernetes/README.md (hunk 5 above)
Comment (silvin-lubecki) [reply kubernetes/README.md]: Sure!

File: cli/command/stack/kubernetes/stack.go (hunk 1 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stack.go]: Right, I'll fix that.

File: cli/command/stack/kubernetes/stack.go (hunk 2 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stack.go]: I'll fix that too.

File: cli/command/stack/kubernetes/stack.go (hunk 1 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stack.go]: I propose to fix this in a follow-up, as we may remove all configmaps and secrets if any error occurs, even while deploying the stack.

File: kubernetes/client/clientset/clientset.go (hunk 6 above)
Comment (silvin-lubecki) [reply kubernetes/client/clientset/clientset.go]: Yes please 😄

File: kubernetes/client/clientset/clientset.go (hunk 7 above)
Comment (silvin-lubecki) [reply kubernetes/client/clientset/clientset.go]: Right

Comment (silvin-lubecki) [reply kubernetes/client/clientset/clientset.go]: Or I may remove it...

File: cli/command/stack/kubernetes/stackclient.go (hunk 3 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stackclient.go]: ok, fixed

File: cli/command/stack/kubernetes/stackclient.go (hunk 4 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stackclient.go]: ok, fixed

**Comment by thaJeztah** [comment]
looks like this needs a rebase

**Comment by silvin-lubecki** [comment]
Done @thaJeztah

**Comment by thaJeztah** [comment]
@silvin-lubecki @vdemeester what's the status on this one? I think there's some outstanding review comments, but perhaps I looked wrong 😅

**Comment by silvin-lubecki** [comment]
I think I addressed all @vdemeester comments, but I'd be pleased to hear your comments @thaJeztah 😄 
Also #903 depends on this PR.

File: cli/command/stack/kubernetes/stackclient.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+func newStackV1Beta2(config *rest.Config, namespace string) (stackClient, error) {
+	client, err := composev1beta2.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &stackV1Beta2{stacks: client.Stacks(namespace)}, nil
+}
+
+func (s *stackV1Beta2) CreateOrUpdate(internalStack stack) error {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stackclient.go]: Question: should we send the `composefile` field or not in that case ? (honestly, I would *not* send it but…)

File: cli/command/stack/kubernetes/loader.go [hunk 9]
```diff
@@ -1,24 +1,33 @@
...
 	composetypes "github.com/docker/cli/cli/compose/types"
-	apiv1beta1 "github.com/docker/cli/kubernetes/compose/v1beta1"
 	yaml "gopkg.in/yaml.v2"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// LoadStack loads a stack from a Compose config, with a given name.
-func LoadStack(name string, cfg composetypes.Config) (*apiv1beta1.Stack, error) {
+func loadStack(name string, cfg composetypes.Config) (stack, error) {
 	res, err := yaml.Marshal(cfg)
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/loader.go]: we only need to do that in `v1beta1` case right ? We shouldn't even "try" to do it all the time.

**Comment by silvin-lubecki** [comment]
@vdemeester PTAL

File: cli/command/stack/kubernetes/stackclient.go [hunk 10]
```diff
@@ -100,6 +103,18 @@ func verify(services corev1.ServiceInterface, stackName string, service string)
 	return nil
 }
 
+func (s *stackV1Beta1) LoadStack(name string, cfg composetypes.Config) (stack, error) {
```
Comment (vdemeester) [inline cli/command/stack/kubernetes/stackclient.go]: nit: not sure if `LoadStack` is the correct name for those too :angel:

Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stackclient.go]: PTAL

**Comment by silvin-lubecki** [comment]
@thaJeztah  PTAL

File: cli/command/stack/kubernetes/stack.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+			continue
+		}
+
+		fileName := path.Base(config.File)
+		content, err := ioutil.ReadFile(config.File)
+		if err != nil {
+			return err
+		}
+
+		if _, err := configMaps.Create(toConfigMap(s.name, name, fileName, content)); err != nil {
```
Comment (thaJeztah) [inline cli/command/stack/kubernetes/stack.go]: Is "create" idempotent, or should we check if a configmap already exists (and is unmodified - if we want to keep the immutable behavior)

File: cli/command/stack/kubernetes/stackclient.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+				return fmt.Errorf("service %s already present in stack named %s", service, key)
+			}
+			return nil
+		}
+		return fmt.Errorf("service %s already present in the cluster", service)
+	}
+	return nil
+}
+
+// stackV1Beta2 implements stackClient interface and talks to compose component v1beta2.
```
Comment (thaJeztah) [inline cli/command/stack/kubernetes/stackclient.go]: nit: tempted to say; move the `stackV1Beta1` and `stackV1Beta2` bits to separate files

File: kubernetes/client/clientset/clientset.go [hunk 13]
```diff
@@ -1,38 +1,40 @@
...
+	composev1beta2 "github.com/docker/cli/kubernetes/client/clientset/typed/compose/v1beta2"
 	glog "github.com/golang/glog"
 	discovery "k8s.io/client-go/discovery"
 	rest "k8s.io/client-go/rest"
 	flowcontrol "k8s.io/client-go/util/flowcontrol"
 )
 
+// Interface defines the methods a compose kube client should have
+// FIXME(vdemeester) is it required ?
 type Interface interface {
```
Comment (thaJeztah) [inline kubernetes/client/clientset/clientset.go]: `Interface` is a terrible name for an interface, lol

File: kubernetes/client/clientset/clientset.go [hunk 14]
```diff
@@ -72,6 +78,7 @@ func NewForConfig(c *rest.Config) (*Clientset, error) {
 // panics if there is an error in the config.
 func NewForConfigOrDie(c *rest.Config) *Clientset {
```
Comment (thaJeztah) [inline kubernetes/client/clientset/clientset.go]: `NewForConfigOrDie()` doesn't seem to be used

File: kubernetes/client/clientset/typed/compose/v1beta2/compose_client.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+	return &ComposeV1beta2Client{client}, nil
+}
+
+// NewForConfigOrDie creates a new ComposeV1beta2Client for the given config and
+// panics if there is an error in the config.
+func NewForConfigOrDie(c *rest.Config) *ComposeV1beta2Client {
```
Comment (thaJeztah) [inline kubernetes/client/clientset/typed/compose/v1beta2/compose_client.go]: `NewForConfigOrDie()` is only used by the wrapper `NewForConfigOrDie()`, which doesn't seem to be used?

File: kubernetes/client/clientset/typed/compose/v1beta2/stack.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+		Resource("stacks").
+		Name(stack.Name).
+		Body(stack).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// UpdateStatus was generated because the type contains a Status member.
+// Add a +genclientstatus=false comment above the type to avoid generating UpdateStatus().
```
Comment (thaJeztah) [inline kubernetes/client/clientset/typed/compose/v1beta2/stack.go]: Looks like `UpdateStatus()` isn't used? should we remove this? (same for the other API version)

File: kubernetes/client/clientset/typed/compose/v1beta1/stack.go [hunk 17]
```diff
@@ -46,44 +45,44 @@ func newStacks(c *ComposeV1beta1Client, namespace string) *stacks {
...
 		Name(stack.Name).
 		Body(stack).
 		Do().
 		Into(result)
-	return
+	return result, err
 }
 
 // UpdateStatus was generated because the type contains a Status member.
 // Add a +genclientstatus=false comment above the type to avoid generating UpdateStatus().
```
Comment (thaJeztah) [inline kubernetes/client/clientset/typed/compose/v1beta1/stack.go]: Looks like `UpdateStatus()` isn't used? should we remove this? (same for the other API version)

File: kubernetes/compose/v1beta1/doc.go [hunk 18]
```diff
@@ -1,11 +1,10 @@
-// Package v1beta1 holds the v1beta1 versions of our stack structures.
```
Comment (thaJeztah) [inline kubernetes/compose/v1beta1/doc.go]: Hm, why was the Package description move further down? The first line is what will be documented as package description

File: kubernetes/compose/v1beta1/stack_test.go [hunk 19]
```diff
@@ -0,0 +1 @@
+package v1beta1
```
Comment (thaJeztah) [inline kubernetes/compose/v1beta1/stack_test.go]: This file is here to get representable coverage?

File: kubernetes/compose/v1beta2/doc.go [hunk 20]
```diff
@@ -0,0 +1,6 @@
+// Api versions allow the api contract for a resource to be changed while keeping
+// backward compatibility by support multiple concurrent versions
+// of the same resource
+
+// Package v1beta2 is the second version of the stack, containing a structured spec
```
Comment (thaJeztah) [inline kubernetes/compose/v1beta2/doc.go]: Same question here for positioning of the Package description (should likely be the first line in this file

File: vendor.conf [hunk 21]
```diff
@@ -83,3 +83,4 @@ k8s.io/client-go kubernetes-1.8.2
 k8s.io/kubernetes v1.8.2
 k8s.io/kube-openapi 61b46af70dfed79c6d24530cd23b41440a7f22a5
 vbom.ml/util 928aaa586d7718c70f4090ddf83f2b34c16fdc8d
+github.com/hashicorp/golang-lru 0a025b7e63adc15a622f29b0b2c4c3848243bbf6
```
Comment (thaJeztah) [inline vendor.conf]: nit: please add a newline

File: cli/command/stack/kubernetes/convert.go [hunk 22]
```diff
@@ -1,41 +1,311 @@
...
+	return m
+}
+
+func fromComposeServiceConfig(s composeTypes.ServiceConfig) v1beta2.ServiceConfig {
+	var userID *int64
+	if s.User != "" {
+		numerical, err := strconv.Atoi(s.User)
+		if err == nil {
+			unixUserID := int64(numerical)
+			userID = &unixUserID
```
[the remark this answers is not in the document]
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/convert.go]: Right, fixed!

File: cli/command/stack/kubernetes/convert.go [hunk 23]
```diff
@@ -1,41 +1,311 @@
...
+			Source: config.Source,
+			Target: config.Target,
+			UID:    config.UID,
+			Mode:   config.Mode,
+		}
+	}
+	return c
+}
+
+func fromComposeHealthcheck(h *composeTypes.HealthCheckConfig) *v1beta2.HealthCheckConfig {
```
[the remark this answers is not in the document]
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/convert.go]: After a second thought, the Disable flag is just not used on Kubernetes compose controller, see https://github.com/docker/cli/pull/899/files#diff-9f8dfaca00498f5f3543480c12fec02aR198 , so I don't know if we really need to error on this...

File: kubernetes/client/clientset/clientset.go (hunk 13 above)
Comment (silvin-lubecki) [reply kubernetes/client/clientset/clientset.go]: Unfortunately it comes from Kubernetes.......

File: kubernetes/client/clientset/typed/compose/v1beta1/stack.go (hunk 17 above)
Comment (silvin-lubecki) [reply kubernetes/client/clientset/typed/compose/v1beta1/stack.go]: Same answer 😄

File: kubernetes/compose/v1beta1/stack_test.go (hunk 19 above)
Comment (silvin-lubecki) [reply kubernetes/compose/v1beta1/stack_test.go]: Yep exactly.

File: kubernetes/compose/v1beta2/doc.go (hunk 20 above)
Comment (silvin-lubecki) [reply kubernetes/compose/v1beta2/doc.go]: Done

File: kubernetes/client/clientset/typed/compose/v1beta2/compose_client.go (hunk 15 above)
Comment (silvin-lubecki) [reply kubernetes/client/clientset/typed/compose/v1beta2/compose_client.go]: Same as above.

File: cli/command/stack/kubernetes/stack.go (hunk 11 above)
Comment (silvin-lubecki) [reply cli/command/stack/kubernetes/stack.go]: Create will fail if it already exists, so I think it's ok?

**Comment by vieux** [comment]
LGTM ping @thaJeztah

File: cli/command/stack/kubernetes/convert.go [hunk 24]
```diff
@@ -1,41 +1,311 @@
...
+			value := matches[3]
+			constraint := &v1beta2.Constraint{
+				Operator: operator,
+				Value:    value,
+			}
+			switch {
+			case key == swarmOs:
+				constraints.OperatingSystem = constraint
+			case key == swarmArch:
+				constraints.Architecture = constraint
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply cli/command/stack/kubernetes/convert.go]: Right, so I was looking at the swarm kit constraints, which take a list for platform;

```go
	// Platforms stores all the platforms that the image can run on.
	// This field is used in the platform filter for scheduling. If empty,
	// then the platform filter is off, meaning there are no scheduling restrictions.
	Platforms []Platform `json:",omitempty"`
```

guess it's not new, so ok for this PR

**Pull request merged**